### PR TITLE
Add BRGEMM API versioning to be compatible with different oneDNN versions

### PR DESCRIPTION
**Pull request by CaoE**
oneDNN v3.6 updated the ukernel APIs of `brgemm` and `brgemm_pack_B`. Considering the upgrade of oneDNN,  ukernel API versioning is needed to be compatible with different oneDNN versions.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by yanbing-j** [comment]
Hi @jgong5 , could you please help review this PR? This PR needs to be merged before we upgrade oneDNN v3.6 into PyTorch.

File: aten/src/ATen/native/CPUBlas.cpp [hunk 1]
```diff
@@ -42,12 +42,18 @@ extern "C" void zaxpy_(int *n, void *a, const void *x, int *incx, void *y, int *
...
 #if AT_MKLDNN_ENABLED()
-#include <oneapi/dnnl/dnnl_version.h>
-#endif // oneDNN
-
-#define ONEDNN_UKERNEL_ENABLED (DNNL_VERSION_MAJOR >=3 && DNNL_VERSION_MINOR >=5)
+#include <ideep.hpp>
+#if (IDEEP_VERSION_MAJOR == 3 && IDEEP_VERSION_MINOR == 5)
+#define ONEDNN_UKERNEL_1
+#elif (IDEEP_VERSION_MAJOR >= 3 && IDEEP_VERSION_MINOR >= 6)
+#define ONEDNN_UKERNEL_2
```
Comment (jgong5) [inline aten/src/ATen/native/CPUBlas.cpp:49]: Suggest to add a brief note here describing the difference between these ukernel versions.

**Review by jgong5 — APPROVED** [review]
Could you please add a PR description on the motivation?

File: aten/src/ATen/native/CPUBlas.cpp (hunk 1 above)
Comment (CaoE) [reply aten/src/ATen/native/CPUBlas.cpp:52]: Added.

**Comment by CaoE** [comment]
@peterbell10 Could you please review this PR ? Thanks.

**Comment by CaoE** [comment]
@ezyang Could you please review this PR ? Thanks.

File: aten/src/ATen/native/CPUBlas.cpp [hunk 2]
```diff
@@ -1003,6 +1013,21 @@ struct GemmHelper {
...
+        N,
+        K,
+        bs,
+        ld_a,
+        ld_b,
+        ld_c,
+        get_dnnl_dtype(dt_a),
+        get_dnnl_dtype(dt_b),
+        get_dnnl_dtype(dt_c));
+    brg.set_add_C(beta > 0);
```
Comment (peterbell10) [inline aten/src/ATen/native/CPUBlas.cpp:1028]: Are alpha and beta not used any more?

Comment (CaoE) [reply aten/src/ATen/native/CPUBlas.cpp:1028]: Yes. oneDNN v3.6 removes the setting of `alpha` and uses `set_add_C` instead of beta.

Comment (peterbell10) [reply aten/src/ATen/native/CPUBlas.cpp:1028]: Then you should remove those arguments from `cpublas::brgemm`.

Comment (CaoE) [reply aten/src/ATen/native/CPUBlas.cpp:1028]: To be compatible with oneDNN version 3.5, can we keep these parameters and add checks for beta and alpha in `cpublas::brgemm` for versions >=3.6 ? For example, `TORCH_CHECK(beta == 0 || beta == 1, ...)`, `TORCH_CHECK(alpha == 1, ...)` .

**Comment by ezyang** [comment]
does not look easy to review, need time

File: aten/src/ATen/native/CPUBlas.cpp (hunk 2 above)
Comment (peterbell10) [reply aten/src/ATen/native/CPUBlas.cpp:1028]: Why not just pass `beta = add_C ? 1 : 0` for the old version, then how you can call`cpublas` doesn't depend on the specific version of oneDNN you have installed.

Comment (CaoE) [reply aten/src/ATen/native/CPUBlas.cpp:1028]: Thanks for your comments. I will remove `beta` and `alpha`.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp [hunk 3]
```diff
@@ -791,10 +791,10 @@ void cpu_flash_attention(
       // Move to the next query
       data_index_step(i, batchSize, j, num_head, k, qSlice);
     }
+    if (need_pack) {
+      cpublas::brgemm_release();
```
Comment (ezyang) [inline aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:795]: How exactly does this work? Now you're repeatedly calling this release inside a parallel_for. Is that really what you want?

File: aten/src/ATen/native/CPUBlas.cpp (hunk 2 above)
Comment (CaoE) [reply aten/src/ATen/native/CPUBlas.cpp]: Removed.

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp (hunk 3 above)
Comment (CaoE) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:793]: Yes, release_hw_context need to be called at the end of the parallel section to guarentee that: use AMX within parallel region, and that the state of the CPU is restored afterwards.

**Review by peterbell10 — APPROVED** [review]

File: aten/src/ATen/native/cpu/FlashAttentionKernel.cpp (hunk 3 above)
Comment (jgong5) [reply aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:793]: > How exactly does this work? Now you're repeatedly calling this release inside a parallel_for. Is that really what you want?

Yes - the state if per core.

**Pull request closed without merging**